### PR TITLE
fix(operator): Fix structured logs common fields

### PR DESCRIPTION
**Pull request by periklis**
**What this PR does / why we need it**:
The Loki Operator is using the `logr.Logger` interface extensively to emit structured logs per controller and handler. In addition the controller-runtime is leveraging the `Builder` factory to link reconciliation errors (i.e. any kind of error that is bubbled up and returned by the controllers' `Reconcile()` method) to the managing controller. However the present implementation makes it hard to link top-bottom logs for a specific combo of a controller, a handler and a LokiStack name due to the following issues:
- The controller-runtime constructs a custom logger with fields extracted from `Builder.For()` and `Builder.Named` methods. This approach works when linking reconciler errors to the orginating controller/LokiStack if and only if the controller is built with `Builder.For(lokiv1.LokiStack{}` as the source type. However this is not the case for controllers like the `AlertingRuleReconciler`, `RecordingRuleReconciler` and more critical for the `CertRotationReconciler` rendering searching for errors per LokiStack impossible.
- Although a logger sink per controller is provided in `main.go` the resulting logger instance (used for info/warning logging/non-reconciliation-error logging) needs to be manually augmented with similar fields (`controller`, `lokistack.name`, `lokistack.namespace`) per controller, handler and incoming request. This manual augmentation happens via `logr.Logger.WithValues()` and has been not consistent.

With the following PR the implementation is streamlined to:
1. Use the same set of fields to identify the controller, the handler (called event), the lokistack (name and namespace) whether the structured log emitted is a reconciler error from controller-runtime or an info/warning log from the Loki Operator code base.
2. Controllers without a `Builder.For(lokiv1.LokiStack{}`) setup will use consistently the same fields for  lokistack (name and namespace)  wherever applicable, i.e. `AlertingRuleReconciler`, `RecordingRuleReconciler` and `RulerConfigReconciler` are manually mimicking the same fields to link annotation error with the target LokiStack.

Examples (for the sake of simplicity modeled with `jq`):

- Searching for a lokistack's events when rotate certificates: 
  `jq -c 'select(.lokistack.name=="lokistack-dev" and .event=="createOrRotateCertificates")'`
- Searching for all controller logs (info/error) for a handler:
  `jq -c 'select(.controller=="lokistack" and .event=="createOrUpdateLokiStack")'`
- Search for all reconciler errors for a single lokistack for a controller:
  `jq -c 'select(.controller=="certrotation" and ._message=="Reconciler error" and .lokistack.name=="lokistack-dev")'`

**Which issue(s) this PR fixes**:
Fixes #<issue number>

**Special notes for your reviewer**:
~This PR leaves out the duplicates cleanup of usage references of `kverrors.Wrap`, `log.Info` and `log.Error` across the code base where the Lokistack identifiers (name, namespace) might be logger under an additional  key (e.g. sometimes `key`, `ref`, `name`, etc.). This is left as future work in a separate PR.~ :heavy_check_mark: 

**Checklist**
- [x] Reviewed the [`CONTRIBUTING.md`](https://github.com/grafana/loki/blob/main/CONTRIBUTING.md) guide (**required**)
- [ ] Documentation added
- [x] Tests updated
- [ ] `CHANGELOG.md` updated
  - [ ] If the change is worth mentioning in the release notes, add `add-to-release-notes` label
- [ ] Changes that require user attention or interaction to upgrade are documented in `docs/sources/setup/upgrade/_index.md`
- [ ] For Helm chart changes bump the Helm chart version in `production/helm/loki/Chart.yaml` and update `production/helm/loki/CHANGELOG.md` and `production/helm/loki/README.md`. [Example PR](https://github.com/grafana/loki/commit/d10549e3ece02120974929894ee333d07755d213)
- [ ] If the change is deprecating or removing a configuration option, update the `deprecated-config.yaml` and `deleted-config.yaml` files respectively in the `tools/deprecated-config-checker` directory. [Example PR](https://github.com/grafana/loki/pull/10840/commits/0d4416a4b03739583349934b96f272fb4f685d15)

### CODE REVIEW

File: operator/internal/handlers/dashboards_delete.go [hunk 1]
```diff
@@ -12,20 +13,29 @@ import (
...
+			"object_name", obj.GetName(),
+			"object_kind", obj.GetObjectKind(),
+		)
+
 		key := client.ObjectKeyFromObject(obj)
 		if err := k.Delete(ctx, obj, &client.DeleteOptions{}); err != nil {
 			if apierrors.IsNotFound(err) {
 				continue
 			}
 			return kverrors.Wrap(err, "failed to delete dashboard", "key", key)
```
Comment (JoaoBraveCoding) [inline operator/internal/handlers/dashboards_delete.go:35]:
```suggestion
			return kverrors.Wrap(err, "failed to delete dashboard", "key", klog.KRef(key.Namespace, key.Name))
```

File: operator/internal/handlers/lokistack_enable_zone_awareness.go [hunk 2]
```diff
@@ -57,14 +62,15 @@ func AnnotatePodWithAvailabilityZone(ctx context.Context, log logr.Logger, c k8s
...
+		return kverrors.Wrap(err, "failed to marshal JSON patch for pod", "pod", podRef, "zone", availabilityZone)
 	}
 
 	if err = c.Patch(ctx, pod, client.RawPatch(types.StrategicMergePatchType, mergePatch)); err != nil && !errors.IsNotFound(err) {
-		ll.Error(err, "could not patch the pod annotations", "name", pod.Name)
-		return kverrors.Wrap(err, "Error patching the pod annotations", "name", pod.Name)
+		return kverrors.Wrap(err, "failed to patch pod with availability zone annotations", "pod", podRef)
 	}
 
+	ll.Info("successfully patched availability zone annotations", "pod", pod.Name)
```
Comment (JoaoBraveCoding) [inline operator/internal/handlers/lokistack_enable_zone_awareness.go:72]: We have already set `pod`, line 22
```suggestion
	ll.Info("successfully patched availability zone annotations")
```

File: operator/internal/handlers/lokistack_enable_zone_awareness.go [hunk 3]
```diff
@@ -18,30 +19,34 @@ import (
 
 func AnnotatePodWithAvailabilityZone(ctx context.Context, log logr.Logger, c k8s.Client, pod *corev1.Pod) error {
 	var err error
-
-	ll := log.WithValues("lokistack-pod-zone-annotation event", "createOrUpdatePodWithLabelPred", "pod", pod.Name)
+	ll := log.WithValues("event", eventAnnotateAvailabilityZoneLabels, "pod", pod.Name)
```
Comment (JoaoBraveCoding) [inline operator/internal/handlers/lokistack_enable_zone_awareness.go:22]: Why not use `podRef` here as well?

File: operator/internal/handlers/handlers.go
Comment (JoaoBraveCoding) [inline operator/internal/handlers/handlers.go:1]: Food for thought, IIUC our logs will benefit from these event values, but if an error occurs in a handler they will not have event values. Should we add `event` values when we throw an error?

File: operator/controllers/loki/controllers.go [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+	RecordingRuleCtrlName       = "recordingrule"
+	RulerConfigCtrlName         = "rulerconfig"
+)
+
+type LogContructorType func(request *reconcile.Request) logr.Logger
+
+func genericLogConstructor(log logr.Logger, name string) LogContructorType {
+	return func(_ *reconcile.Request) logr.Logger {
+		log := log
+		return log.WithValues("controller", name)
```
[the remark this answers is not in the document]
Comment (periklis) [reply operator/controllers/loki/controllers.go:28]: The logr API makes the distinction:
- `log.WithValues`: real key values visible on the messages
- `log.WithName`: defines a new sink only which get logged as field `_component="...-controllers-yourcontroller`

**Review by JoaoBraveCoding — APPROVED** [review]
lgtm

File: operator/controllers/loki/controllers.go [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+)
+
+const (
+	AlertingRuleCtrlName        = "alertingrule"
+	CertRotationCtrlName        = "certrotation"
+	DashboardsCtrlName          = "dashboard"
+	LokiStackCtrlName           = "lokistack"
+	LokiStackZoneLabelsCtrlName = "lokistack-zone-labels"
+	RecordingRuleCtrlName       = "recordingrule"
+	RulerConfigCtrlName         = "rulerconfig"
```
Comment (xperimental) [inline operator/controllers/loki/controllers.go:19]: We usually use prefixes to group similar constants together. Does this make sense here as well? (calling them `ControllerNameAlertingRule`, etc., for example)

File: operator/controllers/loki/controllers.go [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+)
+
+type LogContructorType func(request *reconcile.Request) logr.Logger
```
Comment (xperimental) [inline operator/controllers/loki/controllers.go:22]: Small typo and suggestion for renaming (don't know if we did a similar thing already):

```suggestion
type LogConstructorFunc func(request *reconcile.Request) logr.Logger
```

---

Also this type does not need to be public, as far as I can see...

File: operator/controllers/loki/controllers.go [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+				"controllerKind", "LokiStack",
+				"lokistack", klog.KRef(req.Namespace, req.Name),
+			)
+		}
+
+		return l
+	}
+}
+
+func logWithLokiStackRef(log logr.Logger, req ctrl.Request, name string) logr.Logger {
```
Comment (xperimental) [inline operator/controllers/loki/controllers.go:48]: I wonder why we need this, when we have the `lokiStackLogConstructor` above which should already produce a `lokistack` field. Do we maybe just need to get our logger from the `Context` (there's a helper `ctrl.LoggerFrom(ctx)`)?

File: operator/controllers/loki/controllers.go [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+func lokiStackLogConstructor(log logr.Logger, name string) LogContructorType {
+	return func(req *reconcile.Request) logr.Logger {
+		log := log
+		l := log.WithValues("controller", name)
+
+		if req != nil {
+			l = l.WithValues(
+				"controllerGroup", lokiv1.GroupVersion.Group,
+				"controllerKind", "LokiStack",
+				"lokistack", klog.KRef(req.Namespace, req.Name),
```
Comment (xperimental) [inline operator/controllers/loki/controllers.go:40]: I wonder if we can avoid the extra dependency to `klog` just for getting an `ObjectRef` to produce a string representation of `namespace/name`.

--edit: Oh, wait it's not producing `namespace/name` but an object with `namespace` and `name` attributes in it. Is that what we want?

File: operator/controllers/loki/certrotation_controller.go [hunk 9]
```diff
@@ -55,27 +57,26 @@ func (r *CertRotationReconciler) Reconcile(ctx context.Context, req ctrl.Request
 	}
 
 	checkExpiryAfter := expiryRetryAfter(rt.TargetCertRefresh)
-	r.Log.Info("Checking if LokiStack certificates expired", "name", req.String(), "interval", checkExpiryAfter.String())
+	ll.Info("checking if certificates expired", "interval", checkExpiryAfter.String())
```
Comment (xperimental) [inline operator/controllers/loki/certrotation_controller.go:60]: I recently wondered if we should remove a few of these outputs that happen every reconciliation and only print output if there's anything "noteworthy". Don't know if this should also happen in this PR though.

**Comment by periklis** [comment]
Closing for a later stage to review this

**Pull request closed without merging**